### PR TITLE
cmake: fix the build WITH_DPDK=ON

### PR DESCRIPTION
**Pull request by tchaikov**
* add Findcryptopp.cmake back
  cryptopp support was dropped in #20015, but it's required by
  src/msg/async/dpdk/TCP.h, which `#include <cryptopp/md5.h>`
  so, to fix the FTBFS of WITH_DPDK=ON, we need to bring
  Findcryptopp.cmake back. it was also removed in #20015.
* pass "-march=core2" when building sources which include dpdk headers.
  i was wrong that the headers shipped by distro are generic.
  the headers use the sse instructions for speedup memcpy, see
  /usr/include/x86_64-linux-gnu/dpdk/rte_memcpy.h .
* also, we need to include the arch specific include directory
  for building with dpdk.

Signed-off-by: Kefu Chai <kchai@redhat.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by tchaikov** [comment]
@liu-chunmei

**Comment by cbodley** [comment]
@tchaikov instead of adding cryptopp back, can we use `ceph::crypto::MD5`? i opened a pr that does this at https://github.com/ceph/ceph/pull/23651

File: cmake/modules/Finddpdk.cmake [hunk 1]
```diff
@@ -64,18 +77,37 @@ foreach(c ${components})
...
+  if(NOT TARGET dpdk::cflags)
+     if(CMAKE_SYSTEM_PROCESSOR MATCHES "amd64|x86_64|AMD64")
+      set(rte_cflags "-march=core2")
+    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "arm|ARM")
+      set(rte_cflags "-march=armv7-a")
+    elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
+      set(rte_cflags "-march=armv8-a+crc")
+    endif()
+    if (rte_cflags)
+      add_library(dpdk::cflags INTERFACE IMPORTED)
```
Comment (tchaikov) [inline cmake/modules/Finddpdk.cmake]: should move this out of `if` block.

**Comment by liu-chunmei** [comment]
I still got the build error: 
from /home/chunmei/ceph-seastar/ceph/src/msg/async/Stack.cc:27:
/usr/lib/gcc/x86_64-linux-gnu/7/include/tmmintrin.h: In function ‘void* rte_memcpy_generic(void*, const void*, size_t)’:
/usr/lib/gcc/x86_64-linux-gnu/7/include/tmmintrin.h:185:1: error: inlining failed in call to always_inline ‘__m128i _mm_alignr_epi8(__m128i, __m128i, int)’: target specific option mismatch
 _mm_alignr_epi8(__m128i __X, __m128i __Y, const int __N)
 ^~~~~~~~~~~~~~~

**Comment by tchaikov** [comment]
@liu-chunmei , thank you for testing it.

in the latest change, i refactored `DPDKStack.cc` in 188510e991dcf3ac919ca8551e9c4de255ffb4f5  so it won't include any dpdk headers. also, this change now incorporates @cbodley 's PR at #23651 .

could you give it a try again?

**Comment by liu-chunmei** [comment]
@tchaikov still has the same build error:
/usr/lib/gcc/x86_64-linux-gnu/7/include/tmmintrin.h: In function ‘void* rte_memcpy_generic(void*, const void*, size_t)’:
/usr/lib/gcc/x86_64-linux-gnu/7/include/tmmintrin.h:185:1: error: inlining failed in call to always_inline ‘__m128i _mm_alignr_epi8(__m128i, __m128i, int)’: target specific option mismatch
 _mm_alignr_epi8(__m128i __X, __m128i __Y, const int __N)
 ^~~~~~~~~~~~~~~

**Comment by tchaikov** [comment]
@liu-chunmei could you pastebin the full output ? `_mm_alignr_epi8` is an SSE3 instruction, per https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html, `-march=core2` should enable it.

**Comment by tchaikov** [comment]
@liu-chunmei ping?

**Comment by liu-chunmei** [comment]
@tchaikov please see https://pad.ceph.com/p/dpdk_build_error,  the output file is too big, I just paste a little part, I am working on perfcounter now, so you can ask the others to test again, if the code works on your machine, maybe something wrong with my build environment.

**Comment by tchaikov** [comment]
```
/usr/lib/gcc/x86_64-linux-gnu/7/include/tmmintrin.h:185:1: error: inlining failed in call to always_inline ‘__m128i _mm_alignr_epi8(__m128i, __m128i, int)’: target specific option mismatch
 _mm_alignr_epi8(__m128i __X, __m128i __Y, const int __N)
 ^~~~~~~~~~~~~~~
In file included from /home/chunmei/ceph-seastar/ceph/build/src/dpdk/include/rte_ether.h:21:0,
                 from /home/chunmei/ceph-seastar/ceph/build/src/dpdk/include/rte_ethdev.h:158,
                 from /home/chunmei/ceph-seastar/ceph/src/msg/async/dpdk/DPDK.h:30,
                 from /home/chunmei/ceph-seastar/ceph/src/msg/async/dpdk/DPDKStack.h:24,
                 from /home/chunmei/ceph-seastar/ceph/src/msg/async/Stack.cc:27:
/home/chunmei/ceph-seastar/ceph/build/src/dpdk/include/rte_memcpy.h:793:1534: note: called from here
  MOVEUNALIGNED_LEFT47(dst, src, n, srcofs);
```

thanks @liu-chunmei . i was compiling with the libdpdk-dev shipped from distro. but with the dpdk compiled from source, i am able to reproduce your FTBFS issue.

in the latest change, i refactored `DPDKStack.h` so it won't include `DPDK.h` anymore. 

i just tested the build of `ceph-common` with or without `libdpdk-dev` installed configured `WITH_DPDK=ON` .

**Comment by tchaikov** [comment]
@cbodley @liu-chunmei could you have a quick review?

**Comment by tchaikov** [comment]
thanks @cbodley ! hopefully we won't need it once seastar-osd is in place.

**Comment by cbodley** [comment]
hey @tchaikov, i'm trying to track down some ceph-mon memory leaks that just started showing up in the rgw suite (http://tracker.ceph.com/issues/36033). they complain about a function called `rte_log_init()`, so i'm assuming it's somehow related to dpdk linkage. do you think it may have been caused by this pr?

**Comment by tchaikov** [comment]
@cbodley see https://github.com/ceph/teuthology/pull/1211

**Pull request merged**